horace_core/sqw/@PixelData/PixelData.m

```diff
@@ -327,41 +440,110 @@
     end
 
     function detector_index = get.detector_idx(obj)
-       detector_index = obj.data(obj.FIELD_INDEX_MAP_('detector_idx'), :);
+        obj = obj.load_first_page_if_data_empty_();
+        detector_index = obj.data(obj.FIELD_INDEX_MAP_('detector_idx'), :);
     end
 
     function obj = set.detector_idx(obj, detector_indices)
-       obj.data(obj.FIELD_INDEX_MAP_('detector_idx'), :) = detector_indices;
+        obj.data(obj.FIELD_INDEX_MAP_('detector_idx'), :) = detector_indices;
     end
 
     function detector_index = get.energy_idx(obj)
-       detector_index = obj.data(obj.FIELD_INDEX_MAP_('energy_idx'), :);
+        obj = obj.load_first_page_if_data_empty_();
+        detector_index = obj.data(obj.FIELD_INDEX_MAP_('energy_idx'), :);
     end
 
     function obj = set.energy_idx(obj, energies)
         obj.data(obj.FIELD_INDEX_MAP_('energy_idx'), :) = energies;
-     end
+    end
 
     function signal = get.signal(obj)
-       signal = obj.data(obj.FIELD_INDEX_MAP_('signal'), :);
+        obj = obj.load_first_page_if_data_empty_();
+        signal = obj.data(obj.FIELD_INDEX_MAP_('signal'), :);
     end
 
     function obj = set.signal(obj, signal)
         obj.data(obj.FIELD_INDEX_MAP_('signal'), :) = signal;
-     end
+    end
 
     function variance = get.variance(obj)
-       variance = obj.data(obj.FIELD_INDEX_MAP_('variance'), :);
+        obj = obj.load_first_page_if_data_empty_();
+        variance = obj.data(obj.FIELD_INDEX_MAP_('variance'), :);
     end
 
     function obj = set.variance(obj, variance)
         obj.data(obj.FIELD_INDEX_MAP_('variance'), :) = variance;
-     end
+    end
 
     function num_pix = get.num_pixels(obj)
-        num_pix = size(obj.data, 2);
+        if isempty(obj.f_accessor_)
+            num_pix = size(obj.data, 2);
+        else
+            num_pix = obj.f_accessor_.npixels;
+        end
+    end
+
+    function file_path = get.file_path(obj)
+        file_path = obj.file_path_;
+    end
+
+    function page_size = get.page_size(obj)
+        % The number of pixels that are held in the current page. If no pixels
+        % are held, this returns the number of pixels that will be loaded into
+        % the next page.
+        %
+        % If "num_pixels in file" < "number of pixels that can fit in memory size",
+        % return num_pixels - this avoids ever reading past end of file
+        if isempty(obj.data)
+            num_bytes_in_val = 8;  % pixel data stored in memory as a double
```

Comment:
would it make sense for this (constant) be defined where the pixel array cache is, or to calculate it from the array (`whos('obj.data').bytes / numel(obj.data)`)?